connect-inject/health_check_resource.go

```diff
@@ -0,0 +1,355 @@
+	kubernetesSuccessReasonMsg = "Kubernetes Health Checks Passing"
+
+	// passing/failing strings passed into Consul health check as Status
+	healthCheckPassing  = "passing"
+	healthCheckCritical = "critical"
+)
+
+type HealthCheckResource struct {
+	Log          hclog.Logger
+	Clientset    kubernetes.Interface
+	ClientConfig *api.Config
+
+	// ConsulPort will be 8500/8501 based on TLS enabled
+	ConsulPort string
+
+	// SyncPeriod is the period by which reconcile gets called, default to 1 minute
+	SyncPeriod time.Duration
+
+	lock sync.Mutex
+}
+
+// Run is the long-running runloop for periodically running Reconcile
+// it initially starts a Reconcile phase at startup and then is called again
+// once every SyncPeriod time
+func (h *HealthCheckResource) Run(stopCh <-chan struct{}) {
+	// Start the background watchers
+	h.Reconcile(stopCh)
+
+	reconcileTimer := time.NewTimer(h.SyncPeriod)
+	defer reconcileTimer.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			h.Log.Info("ConsulSyncer quitting")
+			return
+
+		case <-reconcileTimer.C:
+			h.Reconcile(stopCh)
+			reconcileTimer.Reset(h.SyncPeriod)
+		}
+	}
+}
+
+// Delete is not implemented because it is handled by the preStop phase whereby all services
+// related to the pod are deregistered which also deregisters health checks
+func (h *HealthCheckResource) Delete(string) error {
+	return nil
+}
+
+// Informer starts a sharedindex informer which watches and lists corev1.Pod objects
+// which meet the filter of labelInject
+func (h *HealthCheckResource) Informer() cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(
+		// ListWatch takes a List and Watch function which we filter based on label which was injected
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return h.Clientset.CoreV1().Pods(metav1.NamespaceAll).List(ctx.Background(),
+					metav1.ListOptions{LabelSelector: labelInject})
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return h.Clientset.CoreV1().Pods(metav1.NamespaceAll).Watch(ctx.Background(),
+					metav1.ListOptions{LabelSelector: labelInject})
+			},
+		},
+		&corev1.Pod{}, // the target type (Pod)
+		0,             // no resync (period of 0)
+		cache.Indexers{},
+	)
+}
+
+// Upsert determines if an event should be processed and handled accordingly
+// Two primary use cases are handled, new pods will get a new consul TTL health check
+// registered against their respective agent and service, and updates to pods will have
+// this TTL health check updated to reflect the pod status.
+func (h *HealthCheckResource) Upsert(key string, raw interface{}) error {
+	pod := raw.(*corev1.Pod)
+	if !h.shouldProcess(pod) {
+		return nil
+	}
+	// First grab the IDs which we will use for interacting with Consul agent
+	healthCheckID := h.getConsulHealthCheckID(pod)
+	serviceID := h.getConsulServiceID(pod)
+	consulStatus, reason, err := h.getReadyStatusAndReason(pod)
```

Comment:
Rename `consulStatus` to `podStatus`? Initially, I thought it referred to the status of a health check in consul

Reply:
It is actually the formatted (for Consul consumption) status, so consulStatus .. maybe the function name is misleading.

Reply:
IMO it is ok to call it podStatus in this context as we are passing on the podStatus to consul to set as the status for a service instance

Reply:
it is not actually the pod status though, it is a string like "passing" or "critical"  which is only understood by consul

Reply:
probably shorten it to just status in that case?

Reply:
🔥